lib/routes/user/embedded_widgets/bitrefill_widget.dart

```diff
@@ -0,0 +1,92 @@
+import 'dart:async';
+import 'package:ini/ini.dart';
+import 'package:flutter/services.dart';
+import 'dart:convert' as JSON;
+import 'package:flutter/material.dart';
+import 'package:breez/bloc/blocs_provider.dart';
+import 'package:breez/bloc/invoice/invoice_bloc.dart';
+import 'package:flutter_webview_plugin/flutter_webview_plugin.dart';
+import 'package:breez/theme_data.dart' as theme;
+import 'package:breez/widgets/back_button.dart' as backBtn;
+
+class BitrefillPage extends StatefulWidget {
+  final String _title = "Bitrefill";
+
+  @override
+  State<StatefulWidget> createState() {
+    return new BitrefillPageState();
+  }
+}
+
+class BitrefillPageState extends State<BitrefillPage> {
+  final _widgetWebview = new FlutterWebviewPlugin();
+  var _apiKey = "";
+
+  @override
+  void initState() {
+    super.initState();
+    _getApiKey();
+    _widgetWebview.onStateChanged.listen((state) async {
+      if (state.type == WebViewState.finishLoad) {
+        String script =
+            'window.addEventListener("message", receiveMessage, false);' +
+                'function receiveMessage(event) {Android.getPostMessage(event.data);}';
+        _widgetWebview.evalJavascript(script);
+      }
+    });
+
+    _widgetWebview.onDestroy.listen((_) {
+      if (Navigator.canPop(context)) {
+        Navigator.of(context).pop();
+      }
+    });
+  }
+
+  @override void didChangeDependencies(){
+    InvoiceBloc invoiceBloc = AppBlocsProvider.of<InvoiceBloc>(context);
+
+    _widgetWebview.postMessageStream.listen((value) {
+      final order = JSON.jsonDecode(value);
+      invoiceBloc.newLightningLinkSink.add(order['uri']);
+        _widgetWebview.hide();
+        Navigator.of(context).pushNamed('/home');
+      });
+
+    super.didChangeDependencies();
+  }
+
+  @override
+  void dispose() {
+    _widgetWebview.dispose();
+    super.dispose();
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    return new WebviewScaffold(
+      appBar: new AppBar(
+        leading: backBtn.BackButton(),
+        automaticallyImplyLeading: false,
+        iconTheme: theme.appBarIconTheme,
+        textTheme: theme.appBarTextTheme,
+        backgroundColor: Color.fromRGBO(5, 93, 235, 1.0),
+        title: new Text(
+          widget._title,
+          style: theme.appBarTextStyle,
+        ),
+        elevation: 0.0,
+      ),
+      url: 'https://www.bitrefill.com/embed/lightning/?apiKey=$_apiKey&hideQr',
+      withJavascript: true,
+      withZoom: false,
+    );
+  }
+
+  Future _getApiKey() async {
```

Comment:
This still bothers me.. the fact we parse the configuration to get the API key and concat it to the URL here in the UI. This is really business logic that indicates on a missing bloc.
The only thing that we have at this page that is bitrefill specific is the url and the name.
I suggest to do the following:

- Create a model called: SellerModel with the following props: (url, name)
- Create a bloc (very small for now) called: MarketplaceBloc with the a stream called sellersStream. All the config parsing and generating the right url for the UI will be done inside the bloc
- When you show the list of sellers in the Marketplace route use a StreamBuilder to read from the new bloc stream
- The BitRefill route can be changed to a generic "WebSellerRoute" that will get the SellerModel as context and render a webview, listen to post messages and redirect the invoice to the home page as it does now.

I think these changes puts the building blocs in the right place for the near future.

Reply:
I've completed these steps but there are a few static things left.

- How can we read vendors(sellers) api keys in a generic manner?
- How will we initialize vendor list of MarketplaceBloc, where will the vendor data come from?

Reply:
> * How should we read vendors(sellers) api keys in a generic manner?

every vendor will have his config properties prefixed by its name. e.g:
bitrefill.apikey=XXX
We can also use a different config file for the marketplace and show the marketplace UI only if this config file exists.